add hostlist library for encoding/decoding RFC29 hostlists

grondo opened the pull request:
This PR pulls in the old lsd-tools "hostlist" library with some modifications.

The old code was purged of unnecessary cruft and the coding style largely updated to fit the Flux style.

The hostlist API was modernized a bit (see the commit message for details), and the code itself was broken up into a few testable modules with unit tests. 

The new library will be installed along with flux-core as `libflux-hostlist.so` for optional use by apps and framework projects using flux-core.

The API shares some symbol names with the old hostlist code that is used in a few LLNL projects. However, since the new `hostlist.h` is not pulled in automatically via an include of `<flux/core.h>`, and the symbols are kept separately in `libflux-hostlist.so`, I think this will be ok (you have to opt-in to using the new lib)

garlick commented on src/common/libhostlist/hostname.c:
```diff
@@ -0,0 +1,150 @@
+        return NULL;
+
+    if (!(hn->hostname = strdup (hostname))) {
+        free(hn);
```
call `hostname_destroy()` instead to avoid trouncing errno?

garlick commented on src/common/libhostlist/hostname.c:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    hn->suffix = hn->hostname + idx + 1;
+    hn->num = strtoul (hn->suffix, &p, 10);
```
check for `strtoul()` error?

garlick commented on src/common/libhostlist/hostname.c:
```diff
@@ -0,0 +1,150 @@
+    int saved_errno = errno;
+    if (hn) {
+        hn->suffix = NULL;
+        if (hn->hostname)
```
no need to test for non-NULL here or one after.

garlick commented on src/common/libhostlist/hostrange.c:
```diff
@@ -0,0 +1,592 @@
+    return new;
+
+error:
+    free(new);
```
need to preserve errno

garlick commented on src/common/libhostlist/hostrange.c:
```diff
@@ -0,0 +1,592 @@
+    return new;
+
+error:
+    free(new);
```
preserve errno

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+    return new;
+
+fail:
+    free(new);
```
preserve errno

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+            goto error;
+        }
+    }
+    range->lo = strtoul (str, &q, 10);
```
check for `strtoul()` errors?

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+        goto error;
+    }
+
+    range->hi = (p && *p) ? strtoul (p, &q, 10) : range->lo;
```
check for `strtoul()` errors?

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+    return 0;
+
+  error:
+    free(orig);
```
preserve errno

chu11 replied:
flux code style nit: space after function name

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+    } else
+        hr = hostrange_create_single (str);
+
+    hostlist_append_range (hl, hr);
```
check for `hostrange_create()` / `hostrange_create_single()` error
check for `hostlist_append_range()` error

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+{
+    int i;
+    for (i = 0; i < n; i++) {
+        hostlist_append_hr (hl, pfx, rng->lo, rng->hi, rng->width);
```
check for `hostlist_append_hr()` error

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+        for (j = rng->lo; j <= rng->hi; j++) {
+            char host[4096];
+            struct hostrange * hr;
+            snprintf (host, 4096, "%s%0*lu%s", pfx, rng->width, j, sfx);
```
check for `snprintf()` overflow?

garlick commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+            char host[4096];
+            struct hostrange * hr;
+            snprintf (host, 4096, "%s%0*lu%s", pfx, rng->width, j, sfx);
+            hr = hostrange_create_single (host);
+            hostlist_append_range (hl, hr);
```
check for `hostrange_create_single()` failure
check for `hostliust_append_range()` failure

garlick reviewed (approved):
I called out some error propagation issues and then decided that this may not be time well spent right now (yours or mine) due  to the code's legacy/well proven status, so I gave up part way through.  However, may as well post the ones I noticed and you ignore if you want.

The public API cleanup, and the internal split with separate unit tests is a very nice improvement from before IMHO!

grondo commented:
Sorry, I should have mentioned that besides some updates to style and minor return code issues I left the entire hostlist.c codebase untouched *except* where issues were spotted during development of the new testsuite.

chu11 commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+ * next_tok now handles multiple brackets within the same token,
+ * e.g.  node[01-30]-[1-2,6].
+ */
+static char * next_tok(char *sep, char **str)
```
flux code style nit: space after function name

grondo replied:
Yeah, I did try to run `clang-format` to automatically update some of these coding style issues, but gave up after spending about 40 minutes on it.

Since we already used to have the previous version of the hostlist code in flux-core, I didn't want to spend a few hours updating all style and other minor issues in the code, but only fixed what I came across while developing the unit tests.

In the interest of saving time, we should perhaps avoid doing a full code review at this time.

chu11 commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+/* hostrange compare with void * arguments to allow use with
+ * libc qsort()
+ */
+int _cmp(const void *h1, const void *h2)
```
flux code style nit: space after function name

chu11 commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+ *
+ * Assumes hostlist is locked.
+ */
+static int get_bracketed_list(struct hostlist *hl,
```
flux code style nit: space after function name

chu11 commented on src/common/libhostlist/hostlist.c:
```diff
@@ -0,0 +1,1140 @@
+}
+
+/* return true if a bracket is needed for the range at i in hostlist hl */
+static int is_bracket_needed(struct hostlist *hl, int i)
```
flux code style nit: space after function name

chu11 commented on src/common/libhostlist/hostlist.h:
```diff
@@ -0,0 +1,147 @@
+ *
+ *  Note: Caller is responsible for freeing the returned memory.
+ */
+char * hostlist_pop (struct hostlist * hl);
```
nit organizational thing: shouldn't  `hostlist_shift` be listed first b/c it returns first host, while pop returns the last and should be listed second?  I initially thought the "last host" in this comment was a typo.

grondo replied:
Interesting, I never assumed anything about the order that functions are listed in the header file. Is there some standard or idiom of which I'm not aware?

Another thought, now that `hostlist_first()` sets cursor, `hostlist_shift()` could be replaced with `hostlist_first(hl); hostlist_remove_current (hl);`. If we add a `hostlist_last()` then similarly `hostlist_pop()` could be removed.

chu11 replied:
> Interesting, I never assumed anything about the order that functions are listed in the header file. Is there some standard or idiom of which I'm not aware?

No, just a generality.  For example, I would always think `list_first()` should be listed before `list_next()` and `list_head()` before `list_tail()`.

Perhaps part of this was the use of the word `pop()` and the fact this data structure has the word "list" in it.  So I may have subconsciously thought `first host` as well.

> Another thought, now that hostlist_first() sets cursor, hostlist_shift() could be replaced with hostlist_first(hl); hostlist_remove_current (hl);. If we add a hostlist_last() then similarly hostlist_pop() could be removed.

I think this is a good idea.  I think instinctively I thought there should be a "push()" function to go along with the "pop()" function.

grondo replied:
Yeah, I renamed `push` to append for clarity, so it makes sense to try to get rid of `pop` as well.

chu11 commented on src/common/libhostlist/util.c:
```diff
@@ -0,0 +1,59 @@
+    return width > n ? width - n : 0;
+}
+
+int width_equiv(unsigned long n, int *wn, unsigned long m, int *wm)
```
flux code style nit: space after function name

chu11 commented on src/common/libhostlist/util.c:
```diff
@@ -0,0 +1,59 @@
+    if (wn == wm)
+        return 1;
+
+    npad = zero_padded(n, *wn);
```
flux code style nit: space after function name

grondo commented:
Ok, pushed a set of fixups here to address comments thus far:

 * removed `hostlist_pop()` and `hostlist_shift()`, added `hostlist_last()`
 * more coding style updates
 * minor test update

If this looks ok, I'll squash the fixups

grondo commented:
Intermediate fixes squashed.

The most recent push also includes `flux-hostlist.pc` file.

garlick commented:
I'm good with squashing.  Looks good.

chu11 commented:
LGTM too.  As jim says, this code has been vetted over years, we can probably skip many of the nits.

One thing as an aside, do we want to eventually provide a `flux hostlist` command or something equivalent to the `/usr/bin/hostlist` on our systems.  I know that command has come in handy with scripting. Perhaps something to write up in an issue.

And of course ... the python interface?  I believe some Python hostlist implementations are flying around, not sure if we ever made one ourselves.

grondo commented:
Yeah, we should probably wrap the C with a python interface under `flux.hostlist`, then a `flux hostlist` command, if required, would be pretty trivial.

I'll open an issue on the `flux.hostlist` Python bindings and `flux hostlist` command after this is merged.

grondo commented:
Ok, I've set MWP!
